example/temperature.cpp

```diff
@@ -0,0 +1,143 @@
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include <units/origin.h>
+#include <units/physical/si/base/thermodynamic_temperature.h>
+#include <units/physical/si/prefixes.h>
+#include <units/quantity_io.h>
+#include <units/quantity_point.h>
+#include <units/ratio.h>
+#include <iostream>
+
+using namespace units;
+
+namespace si = units::physical::si;
+
+namespace detail {
+
+struct centikelvin : scaled_unit<ratio(1, 100), si::kelvin> {
+};
+
+}  // namespace detail
+
+struct celsius : alias_unit<si::kelvin, basic_symbol_text{"\u00b0C", "`C"}, no_prefix> {
+};
+
+struct celsius_temperature_origin : point_origin<celsius_temperature_origin> {
+  using reference_origin = si::kelvin_temperature_origin;
+  static constexpr auto reference_offset = si::thermodynamic_temperature<::detail::centikelvin, long int>(27315);
+};
+
+template<UnitOf<si::dim_thermodynamic_temperature> U = celsius, QuantityValue Rep = double>
+using celsius_temperature_point = quantity_point<si::dim_thermodynamic_temperature, U, Rep, celsius_temperature_origin>;
+
+struct fahrenheit : named_scaled_unit<fahrenheit, basic_symbol_text{"\u00b0F", "`F"}, no_prefix, ratio(5, 9), si::kelvin> {
+};
+
+namespace detail {
+
+struct millifahrenheit : scaled_unit<ratio(5, 9000), si::kelvin> {
+};
+
+}  // namespace detail
+
+struct fahrenheit_temperature_origin : point_origin<fahrenheit_temperature_origin> {
+  using reference_origin = si::kelvin_temperature_origin;
+  static constexpr auto reference_offset =
+      si::thermodynamic_temperature<::detail::millifahrenheit, long int>(273150 * 9 / 5 + 32000);
+};
+
+template<UnitOf<si::dim_thermodynamic_temperature> U = fahrenheit, QuantityValue Rep = double>
+using fahrenheit_temperature_point =
+    quantity_point<si::dim_thermodynamic_temperature, U, Rep, fahrenheit_temperature_origin>;
+
+
+namespace detail {
+
+template<typename Org>
+concept DerivedPointOrigin = PointOrigin<Org>&& requires
+{
+  typename Org::reference_origin;
+};
+
+template<PointOrigin Org>
+struct reference_origin {
+  using type = typename Org::reference_origin;
+};
+
+template<PointOrigin Org>
+using reference_origin_t =
+    typename std::conditional_t<DerivedPointOrigin<Org>, reference_origin<Org>, std::type_identity<Org>>::type;
+
+template<QuantityPoint P>
+auto offset_to_reference_origin()
+{
+  using Org = typename P::origin;
+  if constexpr (DerivedPointOrigin<Org>) {
+    return Org::reference_offset;
+  } else {
+    return typename P::quantity_type(0);
+  }
+}
```

Comment:
Why this is in a `temperature.cpp` rather than in some common library header?

Reply:
The current version of the PR doesn't really handle temperatures yet, only the concept of the `PointOrigin`. However, the whole point of that concept is of course to make something like temperatures work. So the example here is at the same time a design exploration/validation. But I guess it makes sense to combine those two things into a single PR.